File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int parent[101];

int findParent(int x) {
    if (parent[x] == x) {
        return x;
    } else {
        return parent[x] = findParent(parent[x]);
    }
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
    
    for (int i = 0; i < n; i++) {
        parent[i] = i;
    }
    
    int count = 0;
    priority_queue<pair<int, pair<int, int>>> pq;
    for (vector<int> v : costs) {
        pq.push({-v[2], {v[0], v[1]}});
    }
    while (count != n - 1) {
        int cost = -pq.top().first;
        int x = pq.top().second.first;
        int y = pq.top().second.second;
        pq.pop();
        
        int px = findParent(x);
        int py = findParent(y);
        
        if (px == py) {
            continue;
        }
        
        answer += cost;
        if (parent[px] < parent[py]) {
            parent[py] = px;
        } else {
            parent[px] = py;
        }
        count++;
    }
    return answer;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
 
    static class Edge implements Comparable<Edge> {
        int from;
        int to;
        int cost;
        public Edge(int from, int to, int cost) {
            this.from = from;
            this.to = to;
            this.cost = cost;
        }
        @Override
        public int compareTo(Edge o) {
            return this.cost - o.cost;
        }
    }
    
    static int[] parents;
    
    public int solution(int n, int[][] costs) {
        int answer = 0;
        parents = new int[n];
        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
        
        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < costs.length; i++) {
            edges.add(new Edge(costs[i][0], costs[i][1], costs[i][2]));
        }
        Collections.sort(edges);
        
        int selectedVertices = 0;
        for (Edge edge : edges) {
            if (findParent(edge.from) == findParent(edge.to)) {
                continue;
            }
            selectedVertices++;
            answer += edge.cost;
            union(edge.from, edge.to);
            
            if (selectedVertices == n - 1) {
                break;
            }
        }
        
        return answer;
    }
    
    static int findParent(int target) {
        if (parents[target] == target) {
            return parents[target];
        }
        
        return parents[target] = findParent(parents[target]);
    }
    
    static void union(int a, int b) {
        int parentA = findParent(a);
        int parentB = findParent(b);
        
        if (parents[parentA] < parents[parentB]) {
            parents[parentB] = parentA;
        } else {
            parents[parentA] = parentB;
        }
    }
}